container_images/pytest/main.py

```diff
@@ -0,0 +1,166 @@
+  "--junit-prefix={envname}",
+  # Write a new junit xml file for each interpreter. The filename pattern
+  # is defined in our Prow instance's configuration:
+  #  https://github.com/GoogleCloudPlatform/oss-test-infra/blob/cc1f0cf1ffecc0e3b75664b0d264abc6165276d1/prow/oss/config.yaml#L98
+  "--junit-xml={env:ARTIFACTS}/junit-{envname}.xml",
+]
+
+
+def setup_execution_root(pyproject_toml_file: str) -> str:
+  """Create execution directory and copy project's code to it.
+
+  Tox creates in-directory files, and its execution fails when the
+  source tree is mounted read-only.
+
+  Args:
+    pyproject_toml_file: Absolute path to the project config file. The
+      containing directory will be recursively copied to the execution root.
+
+  Returns:
+    Absolute path to execution root.
+  """
+  assert (os.path.isabs(pyproject_toml_file) and
+          os.path.exists(pyproject_toml_file))
+
+  source_root = os.path.dirname(pyproject_toml_file)
+  exec_root = tempfile.mkdtemp()
+  shutil.copytree(source_root, exec_root, dirs_exist_ok=True)
+  return exec_root
+
+
+def to_tox_version(py_version: str):
+  """Convert `{major}.{minor}` to `py{major}{minor}`"""
+  if re.fullmatch(r"\d.\d{1,2}", py_version):
+    major, minor = py_version.split(".")
+    return "py{major}{minor}".format(major=major, minor=minor)
+  else:
+    sys.exit("Invalid version number: " + py_version)
+
+
+def write_tox_ini():
+  """Read project configuration from pyproject.toml, and write a new tox.ini
+  file.
+
+  The tox.ini file is generated to ensure that tests are run consistently,
+  and that test reports are written to the correct location.
+  """
+  assert os.path.exists("pyproject.toml")
+
+  with open("pyproject.toml") as f:
+    project_cfg = toml.load(f)
+    test_cfg = project_cfg.get("tool", {}).get("gcp-guest-pytest", {})
+
+  # Which interpreters to enable.
+  envlist = [to_tox_version(env) for env in test_cfg.get("envlist", [])]
+
+  # Test-specific dependencies to install.
+  test_deps = test_cfg.get("test-deps", [])
+
+  if not envlist:
+    sys.exit("pyproject.toml must contain a section [tool.gcp-guest-pytest] "
+             'with a key "envlist" and at least one interpreter.')
+
+  config = configparser.ConfigParser()
+  config["tox"] = {
+    "envlist": ", ".join(envlist),
+  }
+
+  config["testenv"] = {
+    "envlogdir": "{env:ARTIFACTS}/tox/{envname}",
+    "deps": "\n\t".join(_common_test_deps + test_deps),
+    "commands": " ".join(_per_interpreter_command),
+  }
+
+  if os.path.exists("tox.ini"):
+    print("Ignoring existing tox.ini file.")
+    os.remove("tox.ini")
+
+  with open("tox.ini", "w") as f:
+    config.write(f)
+
+
+def save_configs(dst: str):
+  print("Saving config files to " + dst)
+  if not os.path.exists(dst):
+    os.mkdir(dst)
+  for fname in ["pyproject.toml", "tox.ini"]:
+    shutil.copy2(fname, dst)
+
+
+def validate_args(args: typing.List[str]):
+  if (len(args) > 1 and args[1].endswith("pyproject.toml") and
```

Comment:
you require that argv[0] is a full path to this file, but then you cd to the directory and expect this file to exist. that means that the filename in argv[0] (as opposed to the dirname) is useless data, as it can't be anything other than `pyproject.toml`

maybe take a directory path instead, as you're already validating the file exists in that path. it also matches our other parameterized test images.

Reply:
updated to use a directory instead of a path to the config file.